PayloadPkg/OsLoader/LoadImage.c

```diff
@@ -527,63 +474,198 @@ GetTraditionalLinux (
-  EFI_HANDLE                 FsHandle;
-  EFI_HANDLE                 HwPartHandle;
-  EFI_HOB_GUID_TYPE          *GuidHob;
-  EXT_BOOT_LOADER_VERSION    *VersionHob;
-
-  FsHandle     = NULL;
-  HwPartHandle = LoadedImage->HwPartHandle;
-  if (BootOption->FsType >= EnumFileSystemMax) {
-    Status = GetIasImageFromRawPartition (BootOption, LoadedImage);
-    return Status;
-  }
+  LOADED_IMAGES_INFO        *LoadedImagesInfo;
+  LOADED_IMAGE              *LoadedImage;
+  BOOT_IMAGE                *BootImage;
+  UINT8                      BootFlags;
+  EFI_STATUS                 Status;
+  UINT8                      Index;
 
-  // Load image from file system.
-  DEBUG ((DEBUG_INFO, "Init File system\n"));
-  Status = InitFileSystem (BootOption->SwPart, EnumFileSystemTypeAuto, HwPartHandle, &FsHandle);
-  if (EFI_ERROR (Status)) {
-    DEBUG ((DEBUG_INFO, "No partitions found, Status = %r\n", Status));
-    return Status;
+  ASSERT (OsBootOption != NULL);
+
+  BootImage = OsBootOption->Image;
+  BootFlags = OsBootOption->BootFlags;
+
+  LoadedImagesInfo = (LOADED_IMAGES_INFO *)AllocateZeroPool (sizeof (LOADED_IMAGES_INFO));
+  if (LoadedImagesInfo == NULL) {
+    return EFI_OUT_OF_RESOURCES;
   }
+  LoadedImagesInfo->Signature = LOADED_IMAGES_INFO_SIGNATURE;
 
-  Status = GetIasImageFromFs (FsHandle, BootOption, LoadedImage);
-  if (!EFI_ERROR (Status)) {
-    return Status;
+  for (Index = 0; Index < LoadImageTypeMax; Index++) {
+    if ((Index == LoadImageTypeTrusty) && !(BootFlags & BOOT_FLAGS_TRUSTY)) {
+      continue;
+    }
+    if (Index == LoadImageTypeMisc) {
+      continue;
+    }
+    if ((Index >= LoadImageTypeExtra0) && !(BootFlags & BOOT_FLAGS_EXTRA)) {
+      if (!BootImage[Index].LbaImage.Valid) {
+        continue;
+      }
+    }
+
+    LoadedImage = (LOADED_IMAGE *)AllocateZeroPool (sizeof (LOADED_IMAGE));
+    LoadedImage->HwPartHandle   = HwPartHandle;
+    LoadedImage->LoadImageType  = Index;
+
+    //
+    // Load Boot Image from FS or RAW partition
+    //
+    if (FsHandle != NULL) {
+      Status = GetBootImageFromFs (FsHandle, OsBootOption, LoadedImage);
+    } else {
+      Status = GetBootImageFromRawPartition (OsBootOption, LoadedImage);
+    }
+
+    if (EFI_ERROR (Status)) {
+      // UnloadBootImages () will free all unnecessary Memory
+      break;
+    }
+
+    DEBUG ((DEBUG_INFO, "LoadBootImage ImageType-%d Image\n", Index));
+    LoadedImagesInfo->LoadedImageList[Index] = LoadedImage;
   }
 
-  // Enable tradition linux boot only in debug build
-  GuidHob = GetNextGuidHob (&gBootLoaderVersionGuid, GetHobListPtr ());
-  if (GuidHob != NULL) {
-    VersionHob = (EXT_BOOT_LOADER_VERSION *)GET_GUID_HOB_DATA (GuidHob);
-    if ((VersionHob->Version.ImageVersion.BldDebug != 0) ||
-         !FeaturePcdGet (PcdVerifiedBootEnabled)) {
+  //
+  // Launch Traditional Linux for debugging purpose only
+  //
+  if ((DebugCodeEnabled () == TRUE) || !FeaturePcdGet (PcdVerifiedBootEnabled)) {

```

Comment:
we don't need compare 2 BOOLEAN by removing "== TRUE"

Reply:
Let me remove it.